pkg/lib/cockpit-components-terminal.tsx

```diff
@@ -120,71 +204,96 @@
-        this.terminal.loadAddon(new CanvasAddon());
-        this.connectChannel();
+    mountTerminal(state: TerminalState) {
+        this.terminal = state.terminal;
+        this.terminalRef.current?.appendChild(state.element);
+        this.terminal.open(state.element);
+        state.connectChannel();
 
+        if (this.props.fontSize)
+            this.terminal.options.fontSize = this.props.fontSize;
+
+        if (this.props.cols && this.props.rows) {
+            this.resizeTerminal(this.props.cols, this.props.rows);
+        }
+
+        this.setTerminalTheme(this.props.theme || 'black-theme');
+        this.terminal.focus();
+    }
+
+    unmountTerminal(state: TerminalState) {
+        this.terminalRef.current?.removeChild(state.element);
+        if (!this.props.rows)
+            this.resizeTerminal(80, 1);
+    }
+
+    componentDidMount() {
+        this.mountTerminal(this.terminal_state);
         if (!this.props.rows) {
             window.addEventListener('resize', this.onWindowResize);
             this.onWindowResize();
         }
-        this.setTerminalTheme(this.props.theme || 'black-theme');
-        this.terminal.focus();
     }
 
-    resizeTerminal(cols, rows) {
+    resizeTerminal(cols: number, rows: number) {
         this.terminal.resize(cols, rows);
-        this.props.channel.control({
-            window: {
-                rows,
-                cols
-            }
-        });
+        if (this.terminal_state.channel) {
+            this.terminal_state.channel.control({
+                window: {
+                    rows,
+                    cols
+                }
+            } as cockpit.JsonObject as cockpit.ControlMessage);
+        }
     }
 
-    componentDidUpdate(prevProps, prevState) {
-        if (prevProps.fontSize !== this.props.fontSize) {
+    componentDidUpdate(prevProps: TerminalComponentProps, prevState: TerminalComponentState) {
+        if (this.props.state && prevProps.state !== this.props.state) {
+            cockpit.assert(!this.props.channel);
+            cockpit.assert(prevProps.state);
+            this.unmountTerminal(prevProps.state);
+            this.terminal_state = this.props.state;
+            this.mountTerminal(this.terminal_state);
+            if (!this.props.cols || !this.props.rows)
+                this.resizeTerminal(this.state.cols, this.state.rows);
+        }
+
+        if (this.props.channel && prevProps.channel !== this.props.channel) {
+            cockpit.assert(!this.props.state);
+            this.terminal.reset();
+            this.terminal_state.resetChannel(this.props.channel);
+            this.props.channel?.control({
+                window: {
+                    rows: this.state.rows,
+                    cols: this.state.cols
+                }
+            } as cockpit.JsonObject as cockpit.ControlMessage);
+        }
+
+        if (this.props.fontSize && prevProps.fontSize !== this.props.fontSize) {
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-21680-2a204949-20250814-103842-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)